### PR TITLE
Multi network interface routing

### PR DESCRIPTION
**Pull request by jukkar**
Moving routing support from #980 to a separate PR. This contains also some patches that are already presented in other PRs, but I needed them in order to compile this set. Last 5 patches are used by multi interface support.

### CODE REVIEW

**Review by dbkinder — DISMISSED** [review]
+1 for doc

File: drivers/console/Kconfig.ws [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+config WS_CONSOLE_LINE_BUF_SIZE
+	int "WS console line buffer size"
+	default 128
+	help
+	This option can be used to modify the size of the buffer storing
+	console output line, prior to sending it through the network.
+	Of course an output line can be longer than such size, it just
+	means sending it will start as soon as it reaches this size.
+	It really depends on what type of output is expected.
+	If there is a lot of short lines, then lower this value. If there
```
Comment (dbkinder) [inline drivers/console/Kconfig.ws]: If there are ...

File: samples/net/ws_console/README.rst [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+.. code-block:: console
+
+    $ cd $ZEPHYR_BASE/samples/net/ws_console
+    $ make pristine && make run
+
+The default make BOARD configuration for this sample is ``qemu_x86``.
+
+Connect to the console from you browser using these URLs http://[2001:db8::1]
+or http://192.0.2.1
```
Comment (dbkinder) [inline samples/net/ws_console/README.rst]: ... as configured in :file:`prj.conf`

File: drivers/console/Kconfig.ws [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	are longer lines, then raise this value.
```
Comment (dbkinder) [inline drivers/console/Kconfig.ws]: are many long lines,

File: drivers/console/Kconfig.ws [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+config WS_CONSOLE_LINE_BUF_NUMBERS
+	int "WS console line buffers"
+	default 4
+	help
+	This option can be used to modify the amount of line buffers the
```
Comment (dbkinder) [inline drivers/console/Kconfig.ws]: s/amount/number/

File: drivers/console/Kconfig.ws [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	driver can use. It really depends on how much output is meant to be
+	sent, depending on the system load etc. You can play on both
```
Comment (dbkinder) [inline drivers/console/Kconfig.ws]: s/play on/adjust/

File: drivers/console/Kconfig.ws [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+	WS_CONSOLE_LINE_BUF_SIZE and this current option to get the
```
Comment (dbkinder) [inline drivers/console/Kconfig.ws]: s/this current option/WS_CONSOLE_LINE_BUF_NUMBERS options/

File: drivers/console/Kconfig.ws [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	best possible buffer settings you need.
+
+config WS_CONSOLE_SEND_TIMEOUT
+	int "WS console line send timeout"
+	default 100
+	help
+	This option can be used to modify the duration of the timer that kick
```
Comment (dbkinder) [inline drivers/console/Kconfig.ws]: what's the time unit here? what happens with the timer "kicks in"?  What's the significance of the "line feed" (or is is a newline? Needs a rewrite, maybe:

      This option can be used to set the time limit (in milliseconds) before
      a non-empty line buffer without a linefeed character is sent.

File: drivers/console/Kconfig.ws [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	in when a line buffer is not empty but did not yet meet the line feed.
+
+config WS_CONSOLE_SEND_THRESHOLD
+	int "WS console line send threshold"
+	default 5
+	help
+	This option can be used to modify the minimal amount of a line buffer
+	that can be sent by the WS server when nothing has happened for
+	a little while (see WS_CONSOLE_SEND_TIMEOUT) and when the line
+	buffer did not meet the line feed yet.
```
Comment (dbkinder) [inline drivers/console/Kconfig.ws]: I'm not sure what "did not meet" means here.

**Review by dbkinder — CHANGES_REQUESTED** [review]
clarity edits

File: drivers/console/Kconfig.ws (hunk 1 above)
Comment (pfalcon) [reply drivers/console/Kconfig.ws]: :-(

> This contains also some patches that are already presented in other PRs

It didn't make sense to review WebSocket patches, as these are pulled in by a chance...

@jukkar : You know my solution to the problem - having smaller PRs with easier to review changes, and having quicker turnaround on them ;-). (And yep, I don't know why multi-face routing would depend on websocks...)

Comment (pfalcon) [reply drivers/console/Kconfig.ws]: > and having quicker turnaround on them

I completely forgot that it's Sep 6 and the release still wasn't made...

File: subsys/net/ip/ipv6.c [hunk 9]
```diff
@@ -1547,6 +1546,15 @@ static void nd_reachable_timeout(struct k_work *work)
 		return;
 	}
 
+	if (net_rpl_get_interface() && nbr->iface == net_rpl_get_interface()) {
```
Comment (tbursztyka) [inline subsys/net/ip/ipv6.c]: is it possible to have an nbr without an iface set?

If not, then you could remove the first part of this test.

File: subsys/net/ip/ipv6.c [hunk 10]
```diff
@@ -776,6 +776,19 @@ static struct net_pkt *update_ll_reserve(struct net_pkt *pkt,
 	NET_DBG("Adjust reserve old %d new %d",
 		net_pkt_ll_reserve(pkt), reserve);
 
+	if (0) {
```
Comment (tbursztyka) [inline subsys/net/ip/ipv6.c]: a small comment telling when this could be needed for devs would be nice here.

File: include/net/net_pkt.h [hunk 11]
```diff
@@ -170,6 +174,23 @@ static inline void net_pkt_set_iface(struct net_pkt *pkt, struct net_if *iface)
...
+	return pkt->iface;
+#endif
+}
+
+static inline void net_pkt_set_orig_iface(struct net_pkt *pkt,
+					  struct net_if *iface)
+{
+#if defined(CONFIG_NET_ROUTING)
+	pkt->orig_iface = iface;
+#endif
```
Comment (tbursztyka) [inline include/net/net_pkt.h]: do as usual in this file: put the #ifdef around the whole fuction, and as an #else create a dummy macro that does nothing.

File: subsys/net/ip/ipv6.c [hunk 12]
```diff
@@ -202,8 +202,11 @@ static struct net_nbr *nbr_lookup(struct net_nbr_table *table,
 			continue;
 		}
 
-		if (nbr->iface == iface &&
-		    net_ipv6_addr_cmp(&net_ipv6_nbr_data(nbr)->addr, addr)) {
+		if (iface && nbr->iface != iface) {
```
Comment (tbursztyka) [inline subsys/net/ip/ipv6.c]: same question as the other patch, is it possible at this stage to have an nbr->iface which is null?

File: subsys/net/ip/ipv6.c [hunk 13]
```diff
@@ -1307,14 +1363,62 @@ static enum net_verdict handle_ns_input(struct net_pkt *pkt)
...
-						  &ns_hdr->tgt);
+	if (IS_ENABLED(CONFIG_NET_ROUTING)) {
+		ifaddr = net_if_ipv6_addr_lookup(&ns_hdr->tgt, NULL);
+	} else {
+		ifaddr = net_if_ipv6_addr_lookup_by_iface(net_pkt_iface(pkt),
+							  &ns_hdr->tgt);
+	}
+
 	if (!ifaddr) {
+		if (IS_ENABLED(CONFIG_NET_ROUTING)) {
```
Comment (tbursztyka) [inline subsys/net/ip/ipv6.c]: make all the below code block into an inline function. It becomes hairy with all these if/else imbrication and indentations starts to be messy due to 80 chars limit.

File: subsys/net/ip/ipv6.c [hunk 14]
```diff
@@ -3777,30 +3930,84 @@ enum net_verdict net_ipv6_process_pkt(struct net_pkt *pkt)
 #if defined(CONFIG_NET_ROUTE)
```
Comment (tbursztyka) [inline subsys/net/ip/ipv6.c]: maybe all the logic in this code block would need to be put also in another local function with a relevant name. net_ipv6_process_pkt becomes too long now.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
I agree with @pfalcon here, could you split the patchset in logical sets? like: RPL set, IPv6 set, shell set... (I don't know why websocket appears here actually)

**Comment by jukkar** [comment]
> I agree with @pfalcon here, could you split the patchset in logical sets? like: RPL set, IPv6 set, shell set... (I don't know why websocket appears here actually)

The reason is the mistake I did while rebasing. I will send this one with only relevant changes.

**Comment by jukkar** [comment]
@tbursztyka Note that David's comments to Kconfig.ws could be mostly also applied to telnet console driver Kconfig.telnet as I copied the Kconfig options from there.

File: include/net/net_pkt.h (hunk 11 above)
Comment (jukkar) [reply include/net/net_pkt.h]: Different people give different advice in reviews. Some have commented that they want to have inline functions that do nothing (like in this example), or then use macro that does nothing like you suggest here. I kind of like the dummy inline function better as it forces the compiler to check that compilation is ok. The end result is the same in both cases.

File: subsys/net/ip/ipv6.c (hunk 12 above)
Comment (jukkar) [reply subsys/net/ip/ipv6.c]: Yes, it is possible now. That is why I needed to add this check.

File: subsys/net/ip/ipv6.c (hunk 13 above)
Comment (jukkar) [reply subsys/net/ip/ipv6.c]: This is not that bad actually indention wise. The problem with making this inline is that there are couple goto's here which make it actually cumbersome if this is inlined.

File: include/net/net_pkt.h (hunk 11 above)
Comment (tbursztyka) [reply include/net/net_pkt.h]: Either way, my point was that it should be consistent within this file. So I guess you'll have to change all the places in that file that do not do this way.

File: subsys/net/ip/ipv6.c (hunk 14 above)
Comment (jukkar) [reply subsys/net/ip/ipv6.c]: I will move this new block to new function.

File: include/net/net_pkt.h (hunk 11 above)
Comment (jukkar) [reply include/net/net_pkt.h]: We can do that, but not in this PR.

**Review by tbursztyka — APPROVED** [review]

**Comment by jukkar** [comment]
The patches got already merged via another PR so closing this one.

**Pull request closed without merging**